reco_utils/recommender/fbt/fbt.py

```diff
@@ -0,0 +1,311 @@
+            sim_df
+            .loc[sim_df
+                 [self.col_item] ==
+                 sim_df[f'{self.col_item}_paired']]
+            .drop(columns=[f'{self.col_item}_paired'])
+        )
+
+        self._is_fit = True
+        logger.info("Done training")
+
+    def predict(self, test):
+        """Generate new recommendations using a trained FBT model.
+
+        Parameters
+        ----------
+        test : DataFrame
+            DataFrame of users who need to get recommendations
+        remove_seen (bool): flag to remove items seen in training
+        from recommendation
+
+        Returns
+        -------
+        R : DataFrame
+            DataFrame with up to `num_recos` rows per user in X.
+        """
+        if not self._is_fit:
+            raise ValueError(("fit() must be called before predict()!"))
+
+        self._check_dataframe(test)
+        logger.info("Calculating recommendation scores")
+
+        # To get recommendations, merge test dataset of users with the
+        # item similarity on col_item to get all matches
+        all_test_item_matches_df = test.merge(
+            self._model_df,
+            on=self.col_item,
+            how='left'
+        )
+        # Give user may interact with multiple items A, B, C, a
+        # item D may be recommended as a popular paired choice for
+        # each of A, B, C. Will average such scores
+        all_recommendations_df = (
+            all_test_item_matches_df
+            .groupby([self.col_user,
+                     f'{self.col_item}_paired'])[self.col_score]
+            .mean()
+            .reset_index(drop=False)
+            .rename(columns={f'{self.col_item}_paired': self.col_item})
+        )
+        all_recommendations_df[self.col_item] = (
+            all_recommendations_df[self.col_item].astype('int64')
+        )
+
+        return all_recommendations_df
+
+    def recommend_k_items(self,
+                          test,
+                          remove_seen=False,
+                          train=None,
+                          top_k=None):
+        """Recommend top K items for all users which are in the test set
+
+        Args:
+            test (pd.DataFrame): users to test
+            top_k (int): number of top items to recommend
+            remove_seen (bool): flag to remove recommendations
+            that have been seen by user
+
+        Returns:
+            pd.DataFrame: top k recommendation items for each user
+        """
+
+        test_users = test[[self.col_user]].drop_duplicates()
+        all_recommendations_df = self.predict(test)
+
+        if remove_seen:
+            if train is None:
+                raise ValueError("Please provide the training data to "
+                                 "remove seen items!")
+            # copy the DataFrame to avoid modification of the input
+            select_columns = [self.col_user, self.col_item]
+            temp_df = train[select_columns].copy()
```

Comment:
same question here, can we avoid copying the data?

Reply:
Similar setup here in the sar file. Is there some other rationale that the copy operation happens here?

https://github.com/microsoft/recommenders/blob/efaa3d7742183dee0846877e2dc64977098e1977/reco_utils/recommender/sar/sar_singlenode.py#L224

Reply:
removed the copy operation as I'm not changing the input in the function.